---
title: "gencode_team_project"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document:
    latex_engine: xelatex
---

## Loading and setting up required data

```{r}
#wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.basic.annotation.gtf.gz

library(tidyverse)
library(readr)

d = read_delim('gencode.v31.basic.annotation.gtf.gz', 
               delim='\t', skip = 5, progress = F, 
               col_names = F)
# skip : description pass
# col_name -> no columns. so F.

cols = c('chrom', 'source', 'feature_type', 'start', 'end', 'score', 'strand', 'phase', 'info')

d = read_delim('gencode.v31.basic.annotation.gtf.gz', 
               delim='\t', skip = 5, 
               progress = F,
               col_names = cols)
```


## 1. Counting the number of genes for each chromosome in GENCODE

In the ```feature_type``` column of raw ```d``` data, there are various rows containing ```gene```, which contains obviously annotations for genes. We might want to subset those rows from the data frame. 

```{r}
table(d$feature_type)
```


```{r}
d1 = filter(d, feature_type == 'gene')
table(d1$chrom)
```

We've subsetted only the ```gene``` row. Now we can count the number of genes for each chromosome in this data.

##2. Counting the number of genes for each gene biotype per chromosome

Gene biotype data is stored in the ```info``` column. To count the number of genes for each gene biotype, we should extract the gene biotype information from the ```info``` column data and store it in a new column. For doing this work, we can use the ```do.call``` function:

```{r}
d1$gene_biotype <- as.character(do.call(rbind.data.frame, strsplit(d1$info, 'gene_type\\s+"'))[[2]])

d1$gene_biotype <- as.character(do.call(rbind.data.frame, strsplit( d1$gene_biotype, '\\"'))[[1]])
```

Now let's check whether the code did work well.

```{r}
head(d1$gene_biotype)
```



 


