---
title: "gencode_team_project"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document:
    latex_engine: xelatex
---


### Loading and setting up required data

```{r}
#wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_31/gencode.v31.basic.annotation.gtf.gz

library(tidyverse)
library(readr)

d = read_delim('gencode.v31.basic.annotation.gtf.gz', 
               delim='\t', skip = 5, progress = F, 
               col_names = F)
# skip : description pass
# col_name -> no columns. so F.

cols = c('chrom', 'source', 'feature_type', 'start', 'end', 'score', 'strand', 'phase', 'info')

d = read_delim('gencode.v31.basic.annotation.gtf.gz', 
               delim='\t', skip = 5, 
               progress = F,
               col_names = cols)
```


### 1. Counting the number of genes for each chromosome in GENCODE

In the ```feature_type``` column of raw ```d``` data, there are various rows containing ```gene```, which contains obviously annotations for genes. We might want to subset those rows from the data frame. 

```{r}
table(d$feature_type)
```


```{r}
d1 = filter(d, feature_type == 'gene')
table(d1$chrom)
```

We've subsetted only the ```gene``` row. Now we can count the number of genes for each chromosome in this data.

### 2. Counting the number of genes for each gene biotype per chromosome

Gene biotype data is stored in the ```info``` column. To count the number of genes for each gene biotype, we should extract the gene biotype information from the ```info``` column data and store it in a new column. For doing this work, we can use the ```do.call``` function:

```{r}
d1$gene_biotype <- as.character(do.call(rbind.data.frame, strsplit(d1$info, 'gene_type\\s+"'))[[2]])

d1$gene_biotype <- as.character(do.call(rbind.data.frame, strsplit( d1$gene_biotype, '\\"'))[[1]])
```

Now let's check whether the code did work well.

```{r}
head(d1$gene_biotype)
table(d1$gene_biotype)
```

### 3. Identifying length of gene by type of chromosome and gene biotype
 
Let's store the gene length information in a new colum ```gene_length```. The length can be calculated by ```end``` and ```start``` data. 
 
```{r}
d1$gene_length <- (d1$end - d1$start)
d2 <- select(d1, chrom, gene_biotype, gene_length)
```

Grouped the datas by gene biotype and chrom data for further processing:

```{r}
d2_type <- d2 %>% group_by(gene_biotype) %>% summarize(mean_length = mean(gene_length), sd_length = sd(gene_length)) %>% arrange(mean_length)

d2_chrom <- d2 %>% group_by(chrom) %>% summarize(mean_length = mean(gene_length), sd_length = sd(gene_length)) %>% arrange(mean_length)
```

quantile information : heights %>% filter(sex == "Female") %>% summarize(range = quantile(height, c(0, 0.5, 1)))

### 4. Data visualization

```{r}
library(ggplot2)

d2_chrom %>% arrange(chrom)

ggplot(d2_chrom, aes(x = mean_length, y = chrom)) +
  geom_segment(aes(yend = chrom),xend = 0) +
  geom_point(size = 2, color = "orange") +
  theme_bw(base_family = "AppleGothic") +
  theme(panel.grid.major.y = element_blank())
```

```{r}
d2_type %>% arrange(gene_biotype)

ggplot(d2_type, aes(x = mean_length,y = gene_biotype)) +
  geom_segment(aes(yend = gene_biotype),xend = 0) +
  geom_point(size = 2, color = "orange") + 
  theme_bw(base_family = "AppleGothic") +
  theme(panel.grid.major.y = element_blank())
```

### Exercise

1. What is the average length of human genes?

```{r}
mean(d2$gene_length)
```

2. Is the distribution of gene length differed by autosomal and sex chromosomes? Please calculate the quantiles (0%, 25%, 50%, 75%, 100%) of the gene length for each group.

```{r}
my_summary <- function(d2){
  x <- quantile(d2$gene_length, c(0, 0.25, 0.5, 0.75, 1))
  data_frame(min = x[1], Q1 = x[2], median = x[3], Q3 = x[4], max = x[5])
}

d2_chrom_quan <- d2 %>% 
 group_by(chrom) %>% 
  do(my_summary(.))

d2_chrom_quan
```

3. Is the distribution of gene length differed by gene biotype? Please calculate the quantiles (0%, 25%, 50%, 75%, 100%) of the gene length for each group.

```{r}
my_summary <- function(d2){
  x <- quantile(d2$gene_length, c(0, 0.25, 0.5, 0.75, 1))
  data_frame(min = x[1], Q1 = x[2], median = x[3], Q3 = x[4], max = x[5])
}

d2_biotype_quan <- d2 %>% 
  group_by(gene_biotype) %>% 
  do(my_summary(.))

d2_biotype_quan
```











