---
title: "SCN2A_assignment_101_kim"
output:
  html_document:
    df_print: paged
  html_notebook: default
  pdf_document:
    latex_engine: xelatex 
---

## 1. Intro

### 1.1. Background

 *SCN2A* is a voltage-gated sodium channel gene that encodes the neuronal sodium channel *NaV1.2* and plays a critical role in action potential initiation during early neurodevelopment. The latest study demonstrated that it is loss of function mutations that in *SCN2A* that lead to autism spectrum disorders (ASD), in contrast to gain of function, which leads to infantile seizures ([Ben-Shalom 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5796785/)). 

### 1.2. Data Visualization Topic

 The topic for data visualization is to plot the **distribution of amino acid position per protein domain in SCN2A mutation variants**. 어떤 plot 형식 썼는지, plot 그리는 데 어떤 데이터 사용했는지 서술하기.

## 2. Exploring data

### 2.1. Unboxing dataset

```{r}
library(dslabs)
library(dplyr)
library(stringr)
library(tidyverse)
library(splitstackshape)
library(ggpubr)
library(ggplot2)
```

```{r}
# Download the SCN2A mutations
# wget https://www.dropbox.com/s/nxd8s6dppyd7uml/table.scn2a.vep_20190916.filtered.txt

# Download the column headers for the file
# wget https://www.dropbox.com/s/56nm62vd58v95fq/table.scn2a.vep_20190916.headers.txt
```

```{r}
options(stringsAsFactors = F)
d = read.csv('table.scn2a.vep_20190916.filtered.txt', sep='\t')
```

### 2.2. Manipulating data frame

First, let's find out what column names are used in the ```d``` dataset.

```{r}
# What's the column names for the ```d``` dataset?
colnames(d)
```

The amino acid position information is stored in the ```Protein_position``` column. But there's some rows that used ```-``` to compress the position information. 

```{r}
# Let's check them out!
d[grepl('-', d$Protein_position), ]
```

Fortunately, there is a only 1 difference between the datum linked by ```-```. So the things we need to do is just removing the linking ```-``` from the rows. 

```{r}
# Let's remove the unnecessary character.
d <- cSplit(d, 'Protein_position', '-', 'long')
# Was the ```-``` removed well?
d[grepl('-', d$Protein_position), ]
```

Now, we successfully trimmed up the amino acid postion data. The next step is to set up domain information from the aggregated ```DOMAINS``` column. There are various kinds of domain data in the ```DOMAINS``` column like hmmpanther, Pfam, Gene3D, etc and each row optionally has one or two of them. We will focus on just one data type for convenience. As I googled, hmmpanther and Gene3d lack information about how to interpret them and use them. So I chose the Pfam data to use as domain information for further processes.

```{r}
# subset rows that contains Pfam data in the ```DOMAINS``` column
d_pf = d[grepl('Pfam', d$DOMAINS), ]

# To subset only the Pfam data, split the strings by ```,``` in each row
d_pf <- cSplit(d_pf, 'DOMAINS', ',', 'long')

# Newly create ```d_pf1``` dataframe by extracting rows contain Pfam string
d_pf1 <- d_pf[grep("Pfam", d_pf$DOMAINS), ]

# Strsplit the core Pfam data
d_pf1$DOMAINS <- as.character(d_pf1$DOMAINS)
d_pf1$DOMAINS <- do.call(rbind.data.frame, strsplit(d_pf1$DOMAINS, ":"))[[2]]

# Checking
table(d_pf1$DOMAINS)
```

The data processing step has been roughly completed. Now, let's start visualizing data by making a plot using the trimmed data.

## 3. Data Visualization

```{r}
library(ggridges)
library(ggplot2)
library(dplyr)
library(tidyr)
library(forcats)
```

densityplot으로 일단 그려는 놓았지만 별로 좋은 방법은 아닌듯. 정확히 어디서 얼마나 데이터가 나타나는지 알 수가 없다. 나중에 densityplot을 뒤에 깔아놓을수는 있어도... lolliplot이나 rigidplot(bar형태)로 그리도록 하자. 더 좋은 plot형식이 있는지도 찾아보자. 변수가 800개니까 point를 사용하는것도.

**사용할만한 plot 정리**
violin plot이나 densityplot 바닥에 깔아주기

stripchart -> dotplot/point + geom_jitter (x:domain, y : proteinposition, 90도회전)
histogram이나 ridgelineplot


```{r}
ggplot(d_pf1, aes(x = Protein_position, y = DOMAINS)) +
  geom_density_ridges(scale = 3, alpha = 0.5, aes(fill = DOMAINS)) +
  scale_fill_manual(values = c("#00AFBB", "#E7B800", "#FC4E07")) +
  facet_wrap(Group ~ .)
```

gnomAD group과 non_gnomAD group으로 나누어 비교해본다.

```{r}
d_pf1 = d_pf1[!grepl('Unclear', d_pf1$Group), ]
d_pf1$Group <- d_pf1$Group %>% str_replace_all("ASD/DD|BIS|IEE_Mild/Ataxia|IEE|Other|Schizophrenia", "non_gnomAD")
table(d_pf1$Group)
```

```{r}
ggplot(d_pf1, aes(x = Protein_position, y = DOMAINS)) +
  geom_density_ridges(scale = 2, alpha = 0.5, aes(fill = DOMAINS)) +
  scale_fill_manual(values = c("#00AFBB", "#E7B800", "#FC4E07")) +
  facet_grid(Group ~ .)
```







